src/sources/youtube.rs

```diff
@@ -0,0 +1,170 @@
+        uri: P,
+        lazy: bool,
+    ) -> SongbirdResult<Restartable> {
+        let uri = format!("ytsearch1:{}", uri.as_ref());
+        Restartable::new(YouTubeRestarter { uri }, lazy).await
+    }
+
+    pub async fn ytdl_playlist(uri: &str) -> Option<Vec<String>> {
+        let mut child = Command::new("yt-dlp")
+            .args([uri, "--flat-playlist", "-j"])
+            .stdout(Stdio::piped())
+            .spawn()
+            .unwrap();
+
+        if let Some(stdout) = &mut child.stdout {
+            let reader = BufReader::new(stdout);
+
+            let lines = reader.lines().flatten().map(|line| {
+                let entry: Value = serde_json::from_str(&line).unwrap();
+                entry.get("url").unwrap().as_str().unwrap().to_string()
+            });
+
+            Some(lines.collect())
+        } else {
+            None
+        }
+    }
+}
+
+struct YouTubeRestarter<P>
+where
+    P: AsRef<str> + Send + Sync,
+{
+    uri: P,
+}
+
+#[async_trait]
+impl<P> Restart for YouTubeRestarter<P>
+where
+    P: AsRef<str> + Send + Clone + Sync,
+{
+    async fn call_restart(&mut self, time: Option<Duration>) -> SongbirdResult<Input> {
+        let (yt, metadata) = ytdl(self.uri.as_ref()).await?;
+
+        if let Some(time) = time {
+            let ts = format!("{:.3}", time.as_secs_f64());
+            ffmpeg_yt(yt, metadata, &["-ss", &ts]).await
+        } else {
+            ffmpeg_yt(yt, metadata, &[]).await
+        }
+    }
+
+    async fn lazy_init(&mut self) -> SongbirdResult<(Option<Metadata>, Codec, Container)> {
+        _ytdl_metadata(self.uri.as_ref())
+            .await
+            .map(|m| (Some(m), Codec::FloatPcm, Container::Raw))
+    }
+}
+
+async fn ytdl(uri: &str) -> Result<(Child, Metadata), SongbirdError> {
+    let ytdl_args = [
+        "-j",            // print JSON information for video for metadata
+        "--no-simulate", // ensure video is downloaded regardless of printing
+        "-f",
+        "bestaudio", // select best quality audio-only
+        "-R",
+        "infinite",        // infinite number of download retries
+        "--no-playlist",   // only download the video if URL also has playlist info
+        "--ignore-config", // disable all configuration files for a yt-dlp run
+        uri,
+        "-o",
+        "-", // stream data to terminal
+    ];
+
+    let mut yt = Command::new("yt-dlp")
+        .args(&ytdl_args)
+        .stdin(Stdio::null())
+        .stderr(Stdio::piped())
+        .stdout(Stdio::piped())
+        .spawn()?;
+
+    // this rigmarole is required due to the inner synchronous reading context
+    let stderr = yt.stderr.take();
+    let (returned_stderr, value) = task::spawn_blocking(move || {
+        let mut s = stderr.unwrap();
+        let out: SongbirdResult<Value> = {
+            let mut o_vec = vec![];
+            let mut serde_read = BufReader::new(s.by_ref());
+
+            if let Ok(len) = serde_read.read_until(0xA, &mut o_vec) {
```

Comment:
Replace 0xA by a NEWLINE constant or something so that it is readable.

Reply:
Additionally, can we maybe test `0xD + 0xA` as well? Just to validate that the #88 isn't related to line endings.

Reply:
The newline character is 0xA in both platforms. Windows just has a CR before (0xD). Proof of that is that songbird only checks for 0xA and works on both platforms.